riscv/processor.cc

```diff
@@ -207,9 +207,9 @@ void state_t::reset(processor_t* const proc, reg_t max_isa)
     csrmap[CSR_CYCLE] = std::make_shared<counter_proxy_csr_t>(proc, CSR_CYCLE, mcycle);
   }
   if (xlen == 32) {
-    counter_top_csr_t_p minstreth, mcycleh;
-    csrmap[CSR_MINSTRETH] = minstreth = std::make_shared<counter_top_csr_t>(proc, CSR_MINSTRETH, minstret);
-    csrmap[CSR_MCYCLEH] = mcycleh = std::make_shared<counter_top_csr_t>(proc, CSR_MCYCLEH, mcycle);
+    csr_t_p minstreth, mcycleh;
+    csrmap[CSR_MINSTRETH] = minstreth = std::make_shared<rv32_high_csr_t>(proc, CSR_MINSTRETH, -1, minstret);
+    csrmap[CSR_MCYCLEH] = mcycleh = std::make_shared<rv32_high_csr_t>(proc, CSR_MCYCLEH, -1, mcycle);
```

Comment:
Commit 136121d2fe8d1699527789405b5807ba5097e2ef breaks `mcycleh` and `minstreth`; they both read & write only 0 now. Test `rv32mi-p-csr` from `riscv-tests` shows this misbehavior.

I can't see what's wrong with the code, though.

Reply:
OK. I'll test this later.

Reply:
Sorry. When I running rv32mi-p-csr with following command:
`spike --isa=rv32im /workspace/riscv/riscv-tools/riscv-tests/isa/rv32mi-p-csr`
It seems pass(no fail related message). And I don't find cycleh in the code of isa/rv64si/csr.S which is reused by rv32mi/csr.S in the latest version of riscv-tests(commit 56c2471)

Reply:
My mistake; I have local modifications to the test to cover this. I will open a PR on riscv-tests with these changes.

Reply:
Using [this branch](https://github.com/scottj97/riscv-tests/tree/mcycleh) ([this commit](https://github.com/scottj97/riscv-tests/commit/1d669b9b42e82987101db8692ac1583324625423)) I ran the following (after building riscv-tests):

```
./configure --enable-commitlog --enable-misaligned
make
./spike -l --log-commits --isa=rv32imafdc riscv-tests/isa/rv32mi-p-csr > rv32mi-p-csr.out 2>&1
```

The additions I made to the test are not self-checking. The test "passes" both before and after 136121d2fe8d1699527789405b5807ba5097e2ef, but you can see the difference in the log before and after:
```
--- a/rv32mi-p-csr.out
+++ b/rv32mi-p-csr.out
@@ -161,17 +161,17 @@ core   0: 3 0x80000194 (0xb0202273) x4  0x000000b0
 core   0: 0x80000198 (0x00100193) li      gp, 1
 core   0: 3 0x80000198 (0x00100193) x3  0x00000001
 core   0: 0x8000019c (0xb8019073) csrw    mcycleh, gp
-core   0: 3 0x8000019c (0xb8019073) c2944_mcycleh 0x00000001
+core   0: 3 0x8000019c (0xb8019073) c2944_mcycleh 0x00000000
 core   0: 0x800001a0 (0xb8002273) csrr    tp, mcycleh
-core   0: 3 0x800001a0 (0xb8002273) x4  0x00000001
+core   0: 3 0x800001a0 (0xb8002273) x4  0x00000000
 core   0: 0x800001a4 (0x00118193) addi    gp, gp, 1
```

Reply:
OK. Thanks. I'll test it later.

Reply:
In the wide_counter_csr_t::unlogged_write, the upper 32 bits for written val will be ignored. So the mcycleh and minstreth which reuse the high 32 bits of mcycle and minstret will be read & write only 0.

Reply:
Your change in 58706130eca53189d0f073cf13a09572a9bedd49 is not correct; privileged spec says of RV32 counters (emphasis added):

> When MXLEN=32, reads of the mcycle, minstret, and mhpmcountern CSRs return bits 31–0 of the corresponding counter, and **_writes change only bits 31–0_**; reads of the mcycleh, minstreth, and mhpmcounternh CSRs return bits 63–32 of the corresponding counter, and writes change only bits 63–32.

(It's possible RV32 `mstatus` is broken when hypervisor's implemented...it looks like `csrw mstatus` might write the upper 32 bits (aka `mstatush`) to 0?)

Reply:
Yeah, that is a question. Maybe we should make rv32_high_csr_t to store the high part additionally. or may the unlogged_write of original csr to write only the low part for RV32.

Reply:
If `unlogged_write()` only writes the low part for RV32 (and I think it must) then the `rv32_high_csr_t` can't use `orig->unlogged_write()` to update the upper 32 bis.

I've confirmed that RV32 `mstatus` is broken, and always has been, ever since `mstatush` first had writable bits back when #812 was fixed. Doing `csrw mstatus` will erroneously clear the upper 32 bits.

Reply:
Yes, so I add two virtual function to csr_t:

- write_high_half : update the high half of csr
- write_full: update the whole 64 bits of csr